girder/models/file.py

```diff
@@ -94,18 +95,35 @@ def download(self, file, offset=0, headers=True, endByte=None,
         :type extraParameters: str or None
         """
         if file.get('assetstoreId'):
-            return self.getAssetstoreAdapter(file).downloadFile(
-                file, offset=offset, headers=headers, endByte=endByte,
-                contentDisposition=contentDisposition,
-                extraParameters=extraParameters)
+            try:
+                fileDownload = self.getAssetstoreAdapter(file).downloadFile(
+                    file, offset=offset, headers=headers, endByte=endByte,
+                    contentDisposition=contentDisposition,
+                    extraParameters=extraParameters)
+                if inspect.isgeneratorfunction(fileDownload) or inspect.isgenerator(fileDownload):
+                    def downloadGenerator():
+                        for data in fileDownload():
+                            yield data
+                        if endByte is None or endByte >= file['size']:
+                            events.trigger('download.complete', info={'file': file,
+                                                                      'redirect': False})
+                    return downloadGenerator
+                else:
+                    return fileDownload
```

Comment:
No. From what I understand, that wont ever happen with the current set of assetstores. It's to allow future assetstores to behave differently. In which case we don't yet know if the download actually completed or not.